cmd/dryrun/dryrun.go

```diff
@@ -0,0 +1,537 @@
+		return err
+	}
+
+	if cfgPolicy.Status.ComplianceState != policyv1.Compliant {
+		cmd.SilenceUsage = true
+
+		return ErrNonCompliant
+	}
+
+	return nil
+}
+
+const parentName string = "cfgpol-dryrun-parent"
+
+// readPolicy reads the policy file specified in the command flags, ensures that it is either a
+// ConfigurationPolicy or a Policy with exactly one ConfigurationPolicy template, and returns that
+// ConfigurationPolicy object after overriding the remediationAction to `inform`.
+func (d *DryRunner) readPolicy(cmd *cobra.Command) (*policyv1.ConfigurationPolicy, error) {
+	reader, err := os.Open(d.policyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	policyBytes, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	unstruct := unstructured.Unstructured{}
+
+	if err := k8syaml.UnmarshalStrict(policyBytes, &unstruct.Object); err != nil {
+		return nil, err
+	}
+
+	if unstruct.GetAPIVersion() != "policy.open-cluster-management.io/v1" {
+		return nil, fmt.Errorf("unsupported apiVersion: %v, the input policy must be "+
+			"'policy.open-cluster-management.io/v1'", unstruct.GetAPIVersion())
+	}
+
+	cfgpol := policyv1.ConfigurationPolicy{}
+
+	switch unstruct.GetKind() {
+	case "ConfigurationPolicy":
+		if err := k8syaml.UnmarshalStrict(policyBytes, &cfgpol); err != nil {
+			return nil, fmt.Errorf("could not unmarshal input to a ConfigurationPolicy: %w", err)
+		}
+	case "Policy":
+		tmpls, found, err := unstructured.NestedSlice(unstruct.Object, "spec", "policy-templates")
+		if err != nil {
+			return nil, fmt.Errorf("invalid input Policy: %w", err)
+		}
+
+		if !found {
+			return nil, fmt.Errorf("invalid input Policy: no policy-templates found")
+		}
+
+		cfgPolFound := false
+
+		for i, tmpl := range tmpls {
+			tmplMap, ok := tmpl.(map[string]any)
+			if !ok {
+				continue
+			}
+
+			objDef, ok := tmplMap["objectDefinition"]
+			if !ok {
+				continue
+			}
+
+			objDefMap, ok := objDef.(map[string]any)
+			if !ok {
+				continue
+			}
+
+			if objDefMap["kind"] != "ConfigurationPolicy" {
+				continue
+			}
+
+			if cfgPolFound {
+				cmd.Println("Ignoring additional ConfigurationPolicy in input policy")
+
+				continue
+			}
+
+			cfgPolFound = true
+
+			cfgpolBytes, err := json.Marshal(objDef)
+			if err != nil {
+				return nil, fmt.Errorf("unable to marshal policy template back to JSON")
+			}
+
+			if err := k8syaml.UnmarshalStrict(cfgpolBytes, &cfgpol); err != nil {
+				return nil, fmt.Errorf("could not unmarshal input policy template [%v] to a "+
+					"ConfigurationPolicy: %w", i, err)
+			}
+		}
+
+		if !cfgPolFound {
+			return nil, fmt.Errorf("invalid input Policy: it must contain a ConfigurationPolicy")
+		}
+	default:
+		return nil, fmt.Errorf("unsupported input kind: %v, must be 'Policy' or 'ConfigurationPolicy'",
+			unstruct.GetKind())
+	}
+
+	cfgpol.Spec.RemediationAction = policyv1.Inform
+	cfgpol.Spec.EvaluationInterval = policyv1.EvaluationInterval{
```

Comment:
So this means the custom messages can't have access to the full object right? I suppose this may be necessary to not be "watch".

Reply:
Good point, I forgot about that when I was testing the custom messages and thought it was an issue with the discovery client. I think now that I added this to avoid the discovery client for a different reason, but maybe we can work around the issue in a different way.

Reply:
After investigating, it seems the fake client doesn't return a resource version from a list query, so the watch is unsupported.